sycl/doc/SpecializationConstants.md

```diff
@@ -0,0 +1,282 @@
+- collects and provides \<Symbolic ID\> =\> \<numeric ID\> spec constant information
+  to the sycl-post-link tool
+Particularly, it translates intrinsic calls to the
+`T __sycl_getSpecConstantValue*(const char *symbolic_id)` intrinsic into
+calls to `T __spirv_SpecConstant(int ID, T default_val)` intrinsic known to
+the llvm-spirv translator. Where `ID` is the numeric ID of the corresponding
+spec constant, `default_val` is its default value which will be used if the
+constant is not set at the runtime.
+
+After this pass the sycl-post-link tool will output the
+\<Symbolic ID\> =\> \<numeric ID\> spec constant mapping into a file for later
+attaching this info to the device binary image via the offload wrapper tool as
+a property set:
+
+```cpp
+struct pi_device_binary_struct {
+...
+  // Array of preperty sets; e.g. specialization constants symbol-int ID map is
+  // propagated to runtime with this mechanism.
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+```
+
+SYCL runtime can then load and access info about particular spec constant using
+its name as a key into the appropriate property set (named "SYCL/specialization
+constants").
+
+##### Ahead of time compilation
+With AOT everything is simplified - the `SpecConstants` pass simply replaces
+the `__sycl_getSpecConstantValue` calls with constants - default values of
+the spec constant's type. No maps are generated, and SYCL program can't change
+the value of a spec constant.
+
+#### LLVMIR-SPIRV translator
+
+Given the `__spirv_SpecConstant` intrinsic calls produced by the
+`SpecConstants` pass:
+```cpp
+; Function Attrs: alwaysinline
+define dso_local spir_func i32 @get() local_unnamed_addr #0 {
+  ; args are "ID" and "default value":
+  %1 = tail call spir_func i32 @_Z20__spirv_SpecConstantii(i32 42, i32 0)
+  ret i32 %1
+}
+```
+the translator will generate `OpSpecConstant` SPIRV instructions with proper
+`SpecId` decorations:
+```cpp
+              OpDecorate %i32 SpecId 42 ; ID
+       %i32 = OpSpecConstant %int 0     ; Default value
+         %1 = OpTypeFunction %int
+
+       %get = OpFunction %int None %1
+         %2 = OpLabel
+              OpReturnValue %i32
+              OpFunctionEnd
+```
+
+#### SYCL runtime
+
+For each device binary compiler generates a map \<Symbolic ID\> =\> \<numeric ID\>
+("ID map"). The SYCL runtime imports that map when loading device binaries.
+It also maintains another map \<Spec const symbolic ID\> =\> \<its value\> 
+("value map") per `sycl::program` object. The value map is updated upon
+`program::set_spec_constant<IDType>(val)` calls from the app.
+
+***NOTE**: `IDType` gets translated to the symbolic ID using the integration
+header mechanism, similarly to kernel ID type. The reason why
+`__builtin_unique_stable_name` is not used here is because this code is
+compiled by the host compiler, which can be any C++ 14-compatible compiler
+unaware of the clang-specific built-ins.*
```

Comment:
```suggestion
**_NOTE_**  `IDType` gets translated to the symbolic ID using the integration
header mechanism, similarly to kernel ID type. The reason why
`__builtin_unique_stable_name` is not used here is because this code is
compiled by the host compiler, which can be any C++ 14-compatible compiler
unaware of the clang-specific built-ins.
```

This should have nicer visualization of the note